pkg/agent/multicast/multicast_route.go

```diff
@@ -0,0 +1,422 @@
+	oif, err := c.vifAllocator.getVif(c.gatewayInterface)
+	if err != nil {
+		return err
+	}
+
+	err = c.addMrouteEntry(src, group, inboundVif, []uint16{oif})
+	if err != nil {
+		return err
+	}
+	routeEntry := &inboundMulticastRouteEntry{
+		group: group.String(),
+		src:   src.String(),
+		vif:   inboundVif,
+	}
+	c.inboundRouteCache.Add(routeEntry)
+	return nil
+}
+
+// inboundMulticastRouteEntry encodes the inbound multicast routing entry.
+// e.g. (10.0.0.55,226.94.9.9)           Iif: wlan0      Oifs: antrea-gw0
+// The oif is always gatewayInterface so we do not put it in the struct.
+type inboundMulticastRouteEntry struct {
+	group string
+	src   string
+	vif   uint16
+}
+
+// outboundMulticastRouteEntry encodes the outbound multicast routing entry.
+// e.g. (10.244.1.4,224.1.2.10)          Iif: antrea-gw0 Oifs: wlo1 wlp3s0 zt5u4zjljr
+// The oif is always gatewayInterface so we do not put it in the struct.
+type outboundMulticastRouteEntry struct {
+	group string
+	src   string
+	vifs  []uint16
+}
+
+func getMulticastInboundEntryKey(obj interface{}) (string, error) {
+	entry := obj.(*inboundMulticastRouteEntry)
+	return entry.group + "/" + entry.src + "/" + fmt.Sprint(entry.vif), nil
+}
+
+func getMulticastOutboundEntryKey(obj interface{}) (string, error) {
+	entry := obj.(*outboundMulticastRouteEntry)
+	return entry.group + "/" + entry.src, nil
+}
+
+func inboundGroupIndexFunc(obj interface{}) ([]string, error) {
+	entry, ok := obj.(*inboundMulticastRouteEntry)
+	if !ok {
+		return []string{}, nil
+	}
+	return []string{entry.group}, nil
+}
+
+func outboundGroupIndexFunc(obj interface{}) ([]string, error) {
+	entry, ok := obj.(*outboundMulticastRouteEntry)
+	if !ok {
+		return []string{}, nil
+	}
+	return []string{entry.group}, nil
+}
+
+// getExternalMulticastAddrMap tries to get external multicast interfaces for configuring outbound multicast routes.
+// Currently the implementation is to add transportInterface to the ExternalMulticastInterfaceConfig.
+func (c *MRouteClient) getExternalMulticastAddrMap(externalInterfaceConfigs []config.ExternalMulticastInterfaceConfig, transportIPv4 *net.IPNet) []config.ExternalMulticastInterfaceConfig {
+	for _, config := range externalInterfaceConfigs {
+		if config.IPv4Addr.IP.Equal(transportIPv4.IP) {
+			return externalInterfaceConfigs
+		}
+	}
+	externalInterfaceConfigs = append(externalInterfaceConfigs, config.ExternalMulticastInterfaceConfig{
+		Name:     c.nodeTransportInterface,
+		IPv4Addr: transportIPv4,
+	})
+	return externalInterfaceConfigs
+}
+
+func setVifToInterface(fd int, vif uint16, index int) error {
+	vc := &multicastsyscall.Vifctl{}
+	vc.Vifi = vif
+	vc.Rate_limit = 0
+	vc.Flags = 0
+	vc.Flags |= multicastsyscall.VIFF_USE_IFINDEX
+	vc.Lcl_ifindex = int32(index)
+	return multicastsyscall.SetsockoptVifctl(fd, syscall.IPPROTO_IP, multicastsyscall.MRT_ADD_VIF, vc)
+}
+
+type parsedIGMPMsg struct {
+	imMsgtype uint8
```

Comment:
I don't find the field is used anywhere.

Reply:
removed